torch/distributed/_composable/fsdp/_fsdp_param.py

```diff
@@ -209,6 +209,11 @@ def _init_sharded_param(self, param: nn.Parameter, device: torch.device):
             global_tp_mesh_dim = _mesh_resources.get_parent_mesh_dim(tp_mesh)
             assert global_dp_mesh_dim is not None  # mypy
             assert global_tp_mesh_dim is not None  # mypy
+            # for PP, DP, TP case, dp mesh dim would be 1, tp mesh dim would be 2
+            # DP/TP would only live in the inner most 2-3 dims
+            if dp_global_mesh.ndim > 2:
+                global_dp_mesh_dim = global_dp_mesh_dim - (dp_global_mesh.ndim - 2)
```

Comment:
iiuc this is shifting dim indies from 
[pp=0, dp=1, tp=2] to [dp=0, tp=1] by 
 - assuming TP+DP are the inner dims
 - discarding the earlier dims

We could more strongly validate the assumption, by asserting that global_tp/dp_mesh_dim are both >0 after the subtraction.  but this looks ok to me.

biggest risk i see is on the device-mesh creation side.  since we dont validate the user-given names at that time, users can specify DP/TP/PP in any order and may be confused about the expected order.  It would be more helpful to raise the error at DM creation time rather than inside one of the parallelism ctors, if possible

Reply:
I think we have thrown a bunch of errors when the mesh passed in is not correct in the TP/DP side. I feel one core design decision we made is to make DeviceMesh be parallelism agnostic, because DeviceMesh should not need to know the parallelism details, it's just a higher level abstraction to process group. Even if we want to error sth in the device mesh creation, it's very hard to do so I feel, the reason is that:

* Given that there're so many different parallelism solutions in the entire PyTorch eco-system, we have seen 3-4 versions of FSDP, a couple of versions of DDP, a couple of versions of megatron based TP and PP, we ultimately can't know what parallelism solution user is using. And We don't want to restrict DeviceMesh to only work with PT-D parallelism I think. 

* I also think that it's ultimately the parallelism implementation restrictions that it require things happen in certain order, ultimately they are just different dimension of devices, if we implement the parallelisms in a composable way, the parallelisms should work in any of the device mesh dimensions. For example, user might want to explore inter-node TP and intra-node PP if that's possible from implementation prospective. I feel let's not put the strict parallelism restriction on what's possible on an abstraction that only model the devices.

lmk if the above does not make sense :)